File: MoviesMVVMAPPWithPagingJava/app/src/main/java/com/masterclass/moviesmvvmappwithpagingjava/utils/GridSpace.java
package com.masterclass.moviesmvvmappwithpagingjava.utils;

import android.graphics.Rect;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

public class GridSpace extends RecyclerView.ItemDecoration {
    private int spaceCount;
    private int spacing;
    private boolean includeEdge;

    public GridSpace(int spaceCount, int spacing, boolean includeEdge) {
        this.spaceCount = spaceCount;
        this.spacing = spacing;
        this.includeEdge = includeEdge;
    }

    @Override
    public void getItemOffsets(@NonNull Rect outRect, @NonNull View view, @NonNull RecyclerView parent, @NonNull RecyclerView.State state) {
        //super.getItemOffsets(outRect, view, parent, state);
        int position = parent.getChildAdapterPosition(view);
        int column = position % spaceCount;

        if (includeEdge){
            outRect.left = spacing - column * spacing / spaceCount;
            outRect.right = (column + 1) * spacing / spaceCount;

            if (position < spaceCount){
                outRect.top = spacing;
            }
            outRect.bottom = spacing;

        }else {
            outRect.left = column * spacing / spaceCount;
            outRect.right = spacing - (column + 1) * spacing / spaceCount;

            if (position >= spaceCount){
                outRect.top = spacing;
            }
        }

    }
}

File: MoviesMVVMAPPWithPagingKotlin/app/src/main/java/com/masterclass/moviesmvvmappwithpagingkotlin/utils/GridSpace.kt
package com.masterclass.moviesmvvmappwithpagingkotlin.utils

import android.graphics.Rect
import android.view.View
import androidx.recyclerview.widget.RecyclerView

class GridSpace(private val spaceCount: Int, private val spacing: Int, private val includeEdge: Boolean) :
    RecyclerView.ItemDecoration() {

    override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: RecyclerView.State) {
        val position = parent.getChildAdapterPosition(view)
        val column = position % spaceCount

        if (includeEdge) {
            outRect.left = spacing - column * spacing / spaceCount
            outRect.right = (column + 1) * spacing / spaceCount

            if (position < spaceCount) {
                outRect.top = spacing
            }
            outRect.bottom = spacing

        } else {
            outRect.left = column * spacing / spaceCount
            outRect.right = spacing - (column + 1) * spacing / spaceCount

            if (position >= spaceCount) {
                outRect.top = spacing
            }
        }
    }
}